### PR TITLE
[NFC][SYCL][Reduction] Rethink getReadWriteAccessorToInitializedMem helper

### PR DESCRIPTION
**Pull request by aelovikov-intel**
Previously the helper abstraction consisted of two implicitly tied pieces that had to be used together and used member variable MOutBufPtr to pass some data behind the scene. Change that to a new abstraction withInitializedMem accepting a callable argument to make the data flow explicit and more immediately clear.

This change also allows to change/optimize this piece in one place without the need to modify multiple uses.

### CODE REVIEW

**Review by steffenlarsen — APPROVED** [review]
Small comments, otherwise LGTM.

File: sycl/include/sycl/reduction.hpp [hunk 1]
```diff
@@ -600,30 +600,57 @@ class reduction_impl_algo : public reduction_impl_common<T, BinaryOperation> {
...
-  /// accessor. Otherwise, create global buffer with 'num_elements' initialized
-  /// with identity value and return an accessor to that buffer.
-  template <bool HasFastAtomics = (has_fast_atomics || has_float64_atomics),
-            typename = std::enable_if_t<HasFastAtomics>>
-  auto getReadWriteAccessorToInitializedMem(handler &CGH) {
-    if constexpr (!is_usm) {
-      if (!base::initializeToIdentity())
-        return MRedOut;
-    }
+  template <typename KernelName, typename FuncTy>
```
[the remark this answers is not in the document]
Comment (aelovikov-intel) [reply sycl/include/sycl/reduction.hpp:611]: Done.

File: sycl/include/sycl/reduction.hpp [hunk 2]
```diff
@@ -600,30 +600,57 @@ class reduction_impl_algo : public reduction_impl_common<T, BinaryOperation> {
...
-    for (int i = 0; i < num_elements; ++i) {
-      (*RWReduVal)[i] = base::getIdentity();
+      reduction::withAuxHandler(CGH, [&](handler &CopyHandler) {
+        accessor Mem{*Buf, CopyHandler};
+        if constexpr (is_usm) {
+          // Can't capture whole reduction, copy into distinct variables.
+          bool IsUpdateOfUserVar = !base::initializeToIdentity();
+          auto BOp = base::getBinaryOperation();
+
+          // Due to constexpr captures with non-default host compilers.
```
[the remark this answers is not in the document]
Comment (aelovikov-intel) [reply sycl/include/sycl/reduction.hpp]: One is about creating an explicit variable, the other is about going from constant expression to a variable. Hopefully the update comment eliminates the confusion.

[the remark this answers is not in the document]
Comment (steffenlarsen) [reply sycl/include/sycl/reduction.hpp]: I think so too, thanks!

**Review by steffenlarsen — APPROVED** [review]
LGTM!

**Pull request merged**